crates/metal-lexer/src/token.rs

```diff
@@ -0,0 +1,243 @@
+    BinaryInteger(u64),
+
+    #[regex("0x[0-9abcdefABCDEF](_?[0-9abcdefABCDEF])*", convert_radix::<16>)]
+    HexInteger(u64),
+
+    #[regex(r"(?&float)[eE][+-]?(?&int)", convert)]
+    ScientificFloat(f64),
+
+    #[regex(r"(?&float)", convert)]
+    Float(f64),
+
+    #[regex("(?&string)", trim_quotes::<false>)]
+    String(&'src str),
+
+    #[regex("true|false", convert)]
+    Boolean(bool),
+
+    /* Comparison */
+
+    #[token("<")]
+    LessThan,
+
+    #[token("<=")]
+    LessThanOrEqual,
+
+    #[token(">")]
+    GreaterThan,
+
+    #[token(">=")]
+    GreaterThanOrEqual,
+
+    #[token("==")]
+    Equal,
+
+    #[token("!=")]
+    NotEqual,
+
+    /* Logical */
+
+    #[token("&&")]
+    And,
+
+    #[token("||")]
+    Or,
+
+    #[token("!")]
+    Not,
+
+    /* Bitwise operators */
+
+    #[token("&")]
+    BitwiseAnd,
+
+    #[token("|")]
+    BitwiseOr,
+
+    #[token("~")]
+    BitwiseNot,
+
+    #[token("^")]
+    BitwiseXor,
+
+    #[token("<<")]
+    BitwiseShiftLeft,
+
+    #[token(">>")]
+    BitwiseShiftRight,
+
+    /* Math */
+
+    #[token("+")]
+    Plus,
+
+    #[token("-")]
+    Minus,
+
+    #[token("*")]
+    Multiply,
+
+    #[token("**")]
+    Power,
+
+    #[token("/")]
+    Divide,
+
+    #[token("%")]
+    Remainder,
+
+    /* Assignment */
+
+    #[token("=")]
+    Assignment,
+
+    #[token("+=")]
+    PlusAssignment,
+
+    #[token("-=")]
+    MinusAssignment,
+
+    #[token("*=")]
+    MultiplyAssignment,
+
+    #[token("**=")]
+    PowerAssignment,
+
+    #[token("/=")]
+    DivideAssignment,
+
+    #[token("%=")]
+    RemainderAssignment,
+
+    #[token("&=")]
+    BitwiseAndAssignment,
+
+    #[token("|=")]
+    BitwiseOrAssignment,
+
+    #[token("~=")]
+    BitwiseNotAssignment,
+
+    #[token("^=")]
+    BitwiseXorAssignment,
+
+    #[token("<<=")]
+    BitwiseShiftLeftAssignment,
+
+    #[token(">>=")]
+    BitwiseShiftRightAssignment,
+
+    /* Control flow */
+
+    #[token("if")]
+    If,
+
+    #[token("else")]
+    Else,
+
+    #[token("for")]
+    For,
+
+    #[token("while")]
+    While,
+
+    #[token("break")]
+    Break,
+
+    #[token("continue")]
+    Continue,
+
+    #[token("return")]
+    Return,
+
+    #[token("switch")]
+    Switch,
+
+    /* Keywords */
+
+    #[token("class")]
+    Class,
+
+    #[token("const")]
+    Const,
+
+    #[token("enum")]
+    Enum,
+
+    #[token("let")]
+    Let,
+
+    /* Punctuation */
+
+    #[regex("@")]
+    Annotation,
+
+    #[token("(")]
+    OpeningParenthesis,
+
+    #[token(")")]
+    ClosingParenthesis,
+
+    #[token("[")]
+    OpeningBracket,
+
+    #[token("]")]
+    ClosingBracket,
+
+    #[token("{")]
+    OpeningBrace,
+
+    #[token("}")]
+    ClosingBrace,
+
+    #[token(",")]
+    Comma,
+
+    #[token(";")]
+    Semicolon,
+
+    #[token(".")]
+    Period,
+
+    #[token("..")]
+    Range,
+
+    #[token(":")]
+    Colon,
+
+    /* Specials */
+
+    #[regex("//[^\n]*")]
+    Comment(&'src str),
```

Comment:
does this regex work with multiline comments?

Reply:
No, multiline comments are a pain in logos because of lookahead, they will be implemented when the FIXME in token.rs is completed. Will open an issue when this gets merged